okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientTransport.java

```diff
@@ -319,36 +323,70 @@ private boolean startPendingStreams() {
   @Override
   public void start(Listener listener) {
     this.listener = Preconditions.checkNotNull(listener, "listener");
-    // We set host to null for test.
-    if (host != null) {
-      BufferedSource source;
-      BufferedSink sink;
-      try {
-        socket = new Socket(host, port);
-        if (sslSocketFactory != null) {
-          socket = OkHttpTlsUpgrader.upgrade(
-              sslSocketFactory, socket, authorityHost, port, connectionSpec);
+
+    frameWriter = new AsyncFrameWriter(this, serializingExecutor);
+    outboundFlow = new OutboundFlowController(this, frameWriter);
+
+    // Connecting in the serializingExecutor, so that some stream operations like synStream
+    // will be executed after connected.
+    serializingExecutor.execute(new Runnable() {
+      @Override
+      public void run() {
+        if (isForTest()) {
+          clientFrameHandler = new ClientFrameHandler();
+          executor.execute(clientFrameHandler);
+          connectedCallback.run();
+          frameWriter.setFrameWriter(rawFrameWriter);
+        } else {
+          BufferedSource source;
+          BufferedSink sink;
+          try {
+            socket = new Socket(host, port);
+            if (sslSocketFactory != null) {
+              socket = OkHttpTlsUpgrader.upgrade(
+                  sslSocketFactory, socket, authorityHost, port, connectionSpec);
+            }
+            socket.setTcpNoDelay(true);
+            source = Okio.buffer(Okio.source(socket));
+            sink = Okio.buffer(Okio.sink(socket));
+          } catch (IOException e) {
+            onIoException(e);
+            // TODO(jhump): should we instead notify the listener of shutdown+terminated?
+            // (and probably do all of this work asynchronously instead of in calling thread)
+            throw Status.UNAVAILABLE.withDescription("Failed connecting").withCause(e)
+                .asRuntimeException();
+          }
+          synchronized (lock) {
+            if (stopped) {
```

Comment:
I don't see that. Can you point out where we may have done cleanups? It seems like cleanups happen on the SerializingExecutor within AsyncFrameWriter, which means they are guaranteed to not have happened yet. From what I can tell, it appears things will work if the code within the `if (stopped)` block is removed.